File: scope.go
package ratify

import (
	"fmt"
	"slices"
	"strings"
)

// Canonical scope constants. Use these instead of bare strings.
// Format: domain:resource:action (colon-separated hierarchy)
const (
	// --- Meeting scopes ---
	ScopeMeetingAttend      = "meeting:attend"
	ScopeMeetingSpeak       = "meeting:speak"
	ScopeMeetingVideo       = "meeting:video"
	ScopeMeetingChat        = "meeting:chat"
	ScopeMeetingShareScreen = "meeting:share_screen"
	ScopeMeetingRecord      = "meeting:record" // sensitive

	// --- Communication scopes ---
	ScopeCommsMessageRead   = "comms:message:read"
	ScopeCommsMessageSend   = "comms:message:send"
	ScopeCommsMessageDelete = "comms:message:delete" // sensitive
	ScopeCommsEmailRead     = "comms:email:read"
	ScopeCommsEmailSend     = "comms:email:send"
	ScopeCommsEmailDelete   = "comms:email:delete" // sensitive
	ScopeCommsCalendarRead  = "comms:calendar:read"
	ScopeCommsCalendarWrite = "comms:calendar:write"

	// --- File scopes ---
	ScopeFilesRead  = "files:read"
	ScopeFilesWrite = "files:write" // sensitive

	// --- Identity scopes ---
	ScopeIdentityProve    = "identity:prove"
	ScopeIdentityDelegate = "identity:delegate" // sensitive

	// --- Presence scopes (candidate v1.x — not yet in canonical vocabulary) ---
	// ScopePresenceRepresent is proposed but NOT YET ADDED to validScopes or
	// sensitiveScopes. See docs/ROADMAP.md §"Candidate v1.x additions" for the
	// full design rationale and open questions before this ships.
	//
	// Intended semantics: the agent is authorized to attend and interact as a
	// direct representative of the principal. Distinct from generate:deepfake
	// (content generation) and identity:delegate (key delegation). Covers both
	// non-likeness representatives and full likeness agents (Tavus, HeyGen, etc.).
	// Sensitive — requires explicit human confirmation.
	//
	// Until this ships, platforms needing representation semantics should use the
	// custom: extension pattern: "custom:presence:represent"
	// ScopePresenceRepresent = "presence:represent" // proposed, sensitive

	// --- Transaction scopes (v1, core to the "transaction horizon" thesis) ---
	// ScopeTransactPurchase: buy goods/services on behalf of the principal
	ScopeTransactPurchase = "transact:purchase"
	// ScopeTransactSell: sell goods/services on behalf of the principal
	ScopeTransactSell = "transact:sell"
	// ScopePaymentsSend: initiate an outbound payment
	ScopePaymentsSend = "payments:send"
	// ScopePaymentsReceive: receive / collect a payment on behalf
	ScopePaymentsReceive = "payments:receive"
	// ScopePaymentsAuthorize: (sensitive) authorize movement of funds from an
	// account beyond standard purchase limits. Requires explicit grant.
	ScopePaymentsAuthorize = "payments:authorize"

	// --- Contract scopes ---
	ScopeContractRead = "contract:read"
	// ScopeContractSign: (sensitive) enter into a binding agreement
	ScopeContractSign = "contract:sign"

	// --- Data scopes (for structured application data, distinct from files) ---
	ScopeDataRead = "data:read"
	// ScopeDataWrite: (sensitive) create or modify data records
	ScopeDataWrite = "data:write"
	// ScopeDataDelete: (sensitive) delete data records
	ScopeDataDelete = "data:delete"
	// ScopeDataExport: (sensitive) bulk export — data exfiltration concern
	ScopeDataExport = "data:export"
	// ScopeDataShare: share data with an authorized third party
	ScopeDataShare = "data:share"

	// --- Execute scopes ---
	// ScopeExecuteTool: invoke an external tool / API on the principal's behalf
	// (covers MCP-style tool calls).
	ScopeExecuteTool = "execute:tool"
	// ScopeExecuteCode: (sensitive) execute arbitrary code on the principal's
	// compute resources. Requires explicit grant.
	ScopeExecuteCode = "execute:code"

	// --- Generate scopes (AI content generation on someone's behalf) ---
	// ScopeGenerateContent: generate text / image / audio / video content.
	ScopeGenerateContent = "generate:content"
	// ScopeGenerateDeepfake: (sensitive) generate content specifically intended
	// to imitate a specific real person. Sensitive by policy so that any
	// such generation creates an auditable explicit authorization trail.
	ScopeGenerateDeepfake = "generate:deepfake"

	// --- Physical-world scopes (v1, first-class coverage for embodied agents) ---
	// The entire Ratify design is channel-agnostic: the same cert/bundle/verify
	// semantics that authorize a software agent also authorize a robot, drone,
	// vehicle, or infrastructure controller. The scopes below are the canonical
	// vocabulary for agents that take action in physical space. Location,
	// temporal, and magnitude bounds are expressed as first-class constraints
	// on the DelegationCert (see types.go Constraint); the scopes below state
	// *what* the agent may do, constraints state *where/when/how much*.

	// ScopePhysicalEnter: enter a physical zone (non-sensitive — the real
	// gate is typically the geo_polygon/geo_circle constraint on the cert).
	ScopePhysicalEnter = "physical:enter"
	// ScopePhysicalExit: exit a physical zone.
	ScopePhysicalExit = "physical:exit"
	// ScopePhysicalActuate: (sensitive) activate a physical actuator — valve,
	// lock, door, latch. Anything that moves matter in the world.
	ScopePhysicalActuate = "physical:actuate"
	// ScopePhysicalManipulate: (sensitive) manipulate physical objects
	// (pick-and-place, lift, rotate). Distinct from actuate: manipulation
	// targets objects, actuation targets fixtures.
	ScopePhysicalManipulate = "physical:manipulate"

	// ScopeRobotOperate: operate a robotic platform (power on, hold, idle
	// motion). The umbrella permission for embodied robots; per-action
	// restrictions use scope chain intersection + constraints.
	ScopeRobotOperate = "robot:operate"
	// ScopeRobotMove: autonomous locomotion — the robot may move in space.
	// Pair with a geo_polygon/geo_circle constraint to bound where.
	ScopeRobotMove = "robot:move"
	// ScopeRobotInteract: the robot may interact with humans or objects in
	// its environment (touch, grasp, gesture). Non-sensitive at the scope
	// level; applications requiring higher assurance combine this with
	// explicit physical:manipulate for stronger semantics.
	ScopeRobotInteract = "robot:interact"

	// ScopeDroneFly: (sensitive) operate a drone under active flight.
	// Nearly always paired with geo + altitude + time_window constraints.
	ScopeDroneFly = "drone:fly"
	// ScopeDroneDeliver: conduct a delivery mission.
	ScopeDroneDeliver = "drone:deliver"
	// ScopeDroneCapture: capture imagery / telemetry data during a flight.
	ScopeDroneCapture = "drone:capture"

	// ScopeVehicleOperate: (sensitive) operate a vehicle — cars, trucks,
	// watercraft, aircraft other than drones. Max-speed and geo constraints
	// are strongly recommended.
	ScopeVehicleOperate = "vehicle:operate"
	// ScopeVehicleTransport: transport a named passenger or payload.
	ScopeVehicleTransport = "vehicle:transport"
	// ScopeVehicleCharge: access charging infrastructure / refueling.
	ScopeVehicleCharge = "vehicle:charge"

	// ScopeInfrastructureMonitor: read sensor values and system state from
	// a piece of infrastructure (HVAC, power, access logs). Read-only.
	ScopeInfrastructureMonitor = "infrastructure:monitor"
	// ScopeInfrastructureControl: (sensitive) modify infrastructure state —
	// HVAC setpoints, breaker state, door policy, etc.
	ScopeInfrastructureControl = "infrastructure:control"
	// ScopeInfrastructureAccess: (sensitive) unlock / grant entry to a
	// restricted facility. Pair with geo + time_window constraints.
	ScopeInfrastructureAccess = "infrastructure:access"

	// ScopeActuateValve: (sensitive) generic valve operation.
	ScopeActuateValve = "actuate:valve"
	// ScopeActuateMotor: (sensitive) generic motor / actuator operation.
	ScopeActuateMotor = "actuate:motor"
	// ScopeActuateSwitch: (sensitive) generic switch / relay operation.
	ScopeActuateSwitch = "actuate:switch"
)

// CustomScopePrefix is the extension-pattern prefix for application-specific
// scopes that are not in the canonical vocabulary. Any scope string starting
// with this prefix is accepted by ValidateScopes, passed through ExpandScopes
// unchanged, and treated as non-sensitive unless the application opts in via
// out-of-band policy.
//
// Example: "custom:acme:inventory:read" — Acme Corp's custom scope for their
// inventory system.
const CustomScopePrefix = "custom:"

// sensitiveScopes require explicit human confirmation beyond standard delegation.
// These are NEVER introduced by wildcard expansion.
var sensitiveScopes = map[string]bool{
	ScopeMeetingRecord:         true,
	ScopeCommsMessageDelete:    true,
	ScopeCommsEmailDelete:      true,
	ScopeFilesWrite:            true,
	ScopeIdentityDelegate:      true,
	ScopePaymentsAuthorize:     true,
	ScopeContractSign:          true,
	ScopeDataWrite:             true,
	ScopeDataDelete:            true,
	ScopeDataExport:            true,
	ScopeExecuteCode:           true,
	ScopeGenerateDeepfake:      true,
	ScopePhysicalActuate:       true,
	ScopePhysicalManipulate:    true,
	ScopeDroneFly:              true,
	ScopeVehicleOperate:        true,
	ScopeInfrastructureControl: true,
	ScopeInfrastructureAccess:  true,
	ScopeActuateValve:          true,
	ScopeActuateMotor:          true,
	ScopeActuateSwitch:         true,
}

// validScopes is the complete canonical vocabulary for v1.
var validScopes = map[string]bool{
	ScopeMeetingAttend:         true,
	ScopeMeetingSpeak:          true,
	ScopeMeetingVideo:          true,
	ScopeMeetingChat:           true,
	ScopeMeetingShareScreen:    true,
	ScopeMeetingRecord:         true,
	ScopeCommsMessageRead:      true,
	ScopeCommsMessageSend:      true,
	ScopeCommsMessageDelete:    true,
	ScopeCommsEmailRead:        true,
	ScopeCommsEmailSend:        true,
	ScopeCommsEmailDelete:      true,
	ScopeCommsCalendarRead:     true,
	ScopeCommsCalendarWrite:    true,
	ScopeFilesRead:             true,
	ScopeFilesWrite:            true,
	ScopeIdentityProve:         true,
	ScopeIdentityDelegate:      true,
	ScopeTransactPurchase:      true,
	ScopeTransactSell:          true,
	ScopePaymentsSend:          true,
	ScopePaymentsReceive:       true,
	ScopePaymentsAuthorize:     true,
	ScopeContractRead:          true,
	ScopeContractSign:          true,
	ScopeDataRead:              true,
	ScopeDataWrite:             true,
	ScopeDataDelete:            true,
	ScopeDataExport:            true,
	ScopeDataShare:             true,
	ScopeExecuteTool:           true,
	ScopeExecuteCode:           true,
	ScopeGenerateContent:       true,
	ScopeGenerateDeepfake:      true,
	ScopePhysicalEnter:         true,
	ScopePhysicalExit:          true,
	ScopePhysicalActuate:       true,
	ScopePhysicalManipulate:    true,
	ScopeRobotOperate:          true,
	ScopeRobotMove:             true,
	ScopeRobotInteract:         true,
	ScopeDroneFly:              true,
	ScopeDroneDeliver:          true,
	ScopeDroneCapture:          true,
	ScopeVehicleOperate:        true,
	ScopeVehicleTransport:      true,
	ScopeVehicleCharge:         true,
	ScopeInfrastructureMonitor: true,
	ScopeInfrastructureControl: true,
	ScopeInfrastructureAccess:  true,
	ScopeActuateValve:          true,
	ScopeActuateMotor:          true,
	ScopeActuateSwitch:         true,
}

// scopeWildcards maps wildcard shorthand to constituent non-sensitive scopes.
// Sensitive scopes are NEVER included in wildcards — they must be granted explicitly.
var scopeWildcards = map[string][]string{
	"meeting:*": {
		ScopeMeetingAttend, ScopeMeetingSpeak, ScopeMeetingVideo,
		ScopeMeetingChat, ScopeMeetingShareScreen,
		// meeting:record excluded (sensitive)
	},
	"comms:message:*": {ScopeCommsMessageRead, ScopeCommsMessageSend},
	"comms:email:*":   {ScopeCommsEmailRead, ScopeCommsEmailSend},
	"comms:*": {
		ScopeCommsMessageRead, ScopeCommsMessageSend,
		ScopeCommsEmailRead, ScopeCommsEmailSend,
		ScopeCommsCalendarRead, ScopeCommsCalendarWrite,
	},
	"transact:*": {ScopeTransactPurchase, ScopeTransactSell},
	"payments:*": {ScopePaymentsSend, ScopePaymentsReceive},
	// data:* excludes write/delete/export (all sensitive)
	"data:*": {ScopeDataRead, ScopeDataShare},
	// execute:* excludes code (sensitive)
	"execute:*": {ScopeExecuteTool},
	// generate:* excludes deepfake (sensitive)
	"generate:*": {ScopeGenerateContent},
	// physical:* excludes actuate/manipulate (both sensitive)
	"physical:*": {ScopePhysicalEnter, ScopePhysicalExit},
	// robot:* — none individually sensitive, but high-stakes robot actions
	// should compose with physical:manipulate for explicit consent.
	"robot:*": {ScopeRobotOperate, ScopeRobotMove, ScopeRobotInteract},
	// drone:* excludes fly (sensitive)
	"drone:*": {ScopeDroneDeliver, ScopeDroneCapture},
	// vehicle:* excludes operate (sensitive)
	"vehicle:*": {ScopeVehicleTransport, ScopeVehicleCharge},
	// infrastructure:* excludes control/access (both sensitive)
	"infrastructure:*": {ScopeInfrastructureMonitor},
	// actuate:* — every member is sensitive, so NO wildcard expansion.
	// Every actuate grant must be explicit.
}

// ValidateScopes returns an error if any scope is not in the canonical
// vocabulary, and is not a wildcard, and does not start with CustomScopePrefix.
//
// Custom scopes are allowed for application-specific extensions. See the
// protocol spec §9.4 for the extension rules.
func ValidateScopes(scopes []string) error {
	for _, s := range scopes {
		if isValidScope(s) {
			continue
		}
		return fmt.Errorf("unknown scope %q: not in canonical vocabulary and not a custom: extension", s)
	}
	return nil
}

// isValidScope returns true iff s is canonical, a wildcard, or a custom extension.
func isValidScope(s string) bool {
	if validScopes[s] {
		return true
	}
	if scopeWildcards[s] != nil {
		return true
	}
	if strings.HasPrefix(s, CustomScopePrefix) && len(s) > len(CustomScopePrefix) {
		return true
	}
	return false
}

// ExpandScopes replaces any wildcard scopes with their constituent scopes.
// Deduplicates the result and returns scopes in lexicographic order so that
// callers (and downstream serializers) see deterministic output.
//
// Custom scopes (prefix "custom:") pass through unchanged — they are never
// expanded by wildcards.
func ExpandScopes(scopes []string) []string {
	seen := make(map[string]bool)
	for _, s := range scopes {
		if children, ok := scopeWildcards[s]; ok {
			for _, c := range children {
				seen[c] = true
			}
		} else {
			seen[s] = true
		}
	}
	result := make([]string, 0, len(seen))
	for s := range seen {
		result = append(result, s)
	}
	slices.Sort(result)
	return result
}

// HasScope checks if the granted scope list covers the required scope,
// including wildcard expansion.
func HasScope(granted []string, required string) bool {
	return slices.Contains(ExpandScopes(granted), required)
}

// IntersectScopes returns the set of scopes present in every input list,
// after wildcard expansion. This is the effective grant of a delegation
// chain: an agent can only exercise scopes that every cert in the chain
// conveyed.
//
// Sensitive scopes are never introduced by wildcard expansion, so a
// sensitive scope not present explicitly in every list is filtered out.
// Custom scopes pass through the intersection normally.
//
// Returns nil for an empty input. For a single list, returns its expansion.
func IntersectScopes(lists ...[]string) []string {
	if len(lists) == 0 {
		return nil
	}
	effective := make(map[string]bool)
	for _, s := range ExpandScopes(lists[0]) {
		effective[s] = true
	}
	for i := 1; i < len(lists); i++ {
		next := make(map[string]bool)
		for _, s := range ExpandScopes(lists[i]) {
			next[s] = true
		}
		for s := range effective {
			if !next[s] {
				delete(effective, s)
			}
		}
	}
	result := make([]string, 0, len(effective))
	for s := range effective {
		result = append(result, s)
	}
	slices.Sort(result)
	return result
}

// IsSensitive returns true if the scope requires explicit human confirmation.
// Custom scopes (prefix "custom:") are NOT sensitive by default; applications
// that want sensitive custom scopes should enforce that at the application
// policy layer.
func IsSensitive(scope string) bool {
	return sensitiveScopes[scope]
}

File: sdks/python/src/ratify_protocol/scope.py
"""Canonical scope vocabulary for Ratify Protocol v1.

MUST stay in lock-step with Go's scope.go, TS's scope.ts, and Rust's scope.rs.
"""
from __future__ import annotations


# --- Meeting scopes ---
SCOPE_MEETING_ATTEND = "meeting:attend"
SCOPE_MEETING_SPEAK = "meeting:speak"
SCOPE_MEETING_VIDEO = "meeting:video"
SCOPE_MEETING_CHAT = "meeting:chat"
SCOPE_MEETING_SHARE_SCREEN = "meeting:share_screen"
SCOPE_MEETING_RECORD = "meeting:record"  # sensitive

# --- Communication scopes ---
SCOPE_COMMS_MESSAGE_READ = "comms:message:read"
SCOPE_COMMS_MESSAGE_SEND = "comms:message:send"
SCOPE_COMMS_MESSAGE_DELETE = "comms:message:delete"  # sensitive
SCOPE_COMMS_EMAIL_READ = "comms:email:read"
SCOPE_COMMS_EMAIL_SEND = "comms:email:send"
SCOPE_COMMS_EMAIL_DELETE = "comms:email:delete"  # sensitive
SCOPE_COMMS_CALENDAR_READ = "comms:calendar:read"
SCOPE_COMMS_CALENDAR_WRITE = "comms:calendar:write"

# --- File scopes ---
SCOPE_FILES_READ = "files:read"
SCOPE_FILES_WRITE = "files:write"  # sensitive

# --- Identity scopes ---
SCOPE_IDENTITY_PROVE = "identity:prove"
SCOPE_IDENTITY_DELEGATE = "identity:delegate"  # sensitive

# --- Presence scopes ---
# SCOPE_PRESENCE_REPRESENT: (sensitive) the agent is authorized to attend
# and interact as a direct representative of the principal — other parties
# in the interaction may be interacting with this agent as if it were the
# principal. Covers both non-likeness representatives and full likeness
# agents (Tavus, HeyGen, etc.).
#
# Distinct from generate:deepfake (content generation, not real-time
# representation) and identity:delegate (key delegation). Does NOT imply
# identity:prove — issuers grant both explicitly when both are needed;
# scope lists are literal, with no implication table.
#
# Verifiers accepting this scope are expected to surface the
# representation relationship to the other participants. That disclosure
# is platform policy, not a protocol constraint — see SPEC §9.1.
SCOPE_PRESENCE_REPRESENT = "presence:represent"  # sensitive

# --- Transaction scopes (v1, core to the "transaction horizon" thesis) ---
SCOPE_TRANSACT_PURCHASE = "transact:purchase"
SCOPE_TRANSACT_SELL = "transact:sell"
SCOPE_PAYMENTS_SEND = "payments:send"
SCOPE_PAYMENTS_RECEIVE = "payments:receive"
SCOPE_PAYMENTS_AUTHORIZE = "payments:authorize"  # sensitive

# --- Contract scopes ---
SCOPE_CONTRACT_READ = "contract:read"
SCOPE_CONTRACT_SIGN = "contract:sign"  # sensitive

# --- Data scopes (structured application data, distinct from files) ---
SCOPE_DATA_READ = "data:read"
SCOPE_DATA_WRITE = "data:write"  # sensitive
SCOPE_DATA_DELETE = "data:delete"  # sensitive
SCOPE_DATA_EXPORT = "data:export"  # sensitive — exfiltration concern
SCOPE_DATA_SHARE = "data:share"

# --- Execute scopes ---
SCOPE_EXECUTE_TOOL = "execute:tool"
SCOPE_EXECUTE_CODE = "execute:code"  # sensitive

# --- Generate scopes (AI content generation on someone's behalf) ---
SCOPE_GENERATE_CONTENT = "generate:content"
# Sensitive by policy: any "imitate a real person" generation creates
# an auditable explicit authorization trail.
SCOPE_GENERATE_DEEPFAKE = "generate:deepfake"  # sensitive

# --- Physical-world scopes (v1, first-class coverage for embodied agents) ---
# Ratify is channel-agnostic: same cert/bundle/verify semantics for software
# agents and for robots, drones, vehicles, infrastructure controllers.
# Location / time / speed / amount / rate bounds live in first-class
# Constraint objects on DelegationCert (see types.py).

SCOPE_PHYSICAL_ENTER = "physical:enter"
SCOPE_PHYSICAL_EXIT = "physical:exit"
SCOPE_PHYSICAL_ACTUATE = "physical:actuate"  # sensitive
SCOPE_PHYSICAL_MANIPULATE = "physical:manipulate"  # sensitive

SCOPE_ROBOT_OPERATE = "robot:operate"
SCOPE_ROBOT_MOVE = "robot:move"
SCOPE_ROBOT_INTERACT = "robot:interact"

SCOPE_DRONE_FLY = "drone:fly"  # sensitive
SCOPE_DRONE_DELIVER = "drone:deliver"
SCOPE_DRONE_CAPTURE = "drone:capture"

SCOPE_VEHICLE_OPERATE = "vehicle:operate"  # sensitive
SCOPE_VEHICLE_TRANSPORT = "vehicle:transport"
SCOPE_VEHICLE_CHARGE = "vehicle:charge"

SCOPE_INFRASTRUCTURE_MONITOR = "infrastructure:monitor"
SCOPE_INFRASTRUCTURE_CONTROL = "infrastructure:control"  # sensitive
SCOPE_INFRASTRUCTURE_ACCESS = "infrastructure:access"  # sensitive

SCOPE_ACTUATE_VALVE = "actuate:valve"  # sensitive
SCOPE_ACTUATE_MOTOR = "actuate:motor"  # sensitive
SCOPE_ACTUATE_SWITCH = "actuate:switch"  # sensitive

# --- Extension pattern ---
# Any scope string starting with CUSTOM_SCOPE_PREFIX is accepted by
# validate_scopes, passes through expand_scopes unchanged, and is treated as
# non-sensitive unless the application opts in via out-of-band policy.
#
# Example: "custom:acme:inventory:read"
CUSTOM_SCOPE_PREFIX = "custom:"


_SENSITIVE_SCOPES: frozenset[str] = frozenset({
    SCOPE_MEETING_RECORD,
    SCOPE_COMMS_MESSAGE_DELETE,
    SCOPE_COMMS_EMAIL_DELETE,
    SCOPE_FILES_WRITE,
    SCOPE_IDENTITY_DELEGATE,
    SCOPE_PRESENCE_REPRESENT,
    SCOPE_PAYMENTS_AUTHORIZE,
    SCOPE_CONTRACT_SIGN,
    SCOPE_DATA_WRITE,
    SCOPE_DATA_DELETE,
    SCOPE_DATA_EXPORT,
    SCOPE_EXECUTE_CODE,
    SCOPE_GENERATE_DEEPFAKE,
    SCOPE_PHYSICAL_ACTUATE,
    SCOPE_PHYSICAL_MANIPULATE,
    SCOPE_DRONE_FLY,
    SCOPE_VEHICLE_OPERATE,
    SCOPE_INFRASTRUCTURE_CONTROL,
    SCOPE_INFRASTRUCTURE_ACCESS,
    SCOPE_ACTUATE_VALVE,
    SCOPE_ACTUATE_MOTOR,
    SCOPE_ACTUATE_SWITCH,
})

_VALID_SCOPES: frozenset[str] = frozenset({
    SCOPE_MEETING_ATTEND,
    SCOPE_MEETING_SPEAK,
    SCOPE_MEETING_VIDEO,
    SCOPE_MEETING_CHAT,
    SCOPE_MEETING_SHARE_SCREEN,
    SCOPE_MEETING_RECORD,
    SCOPE_COMMS_MESSAGE_READ,
    SCOPE_COMMS_MESSAGE_SEND,
    SCOPE_COMMS_MESSAGE_DELETE,
    SCOPE_COMMS_EMAIL_READ,
    SCOPE_COMMS_EMAIL_SEND,
    SCOPE_COMMS_EMAIL_DELETE,
    SCOPE_COMMS_CALENDAR_READ,
    SCOPE_COMMS_CALENDAR_WRITE,
    SCOPE_FILES_READ,
    SCOPE_FILES_WRITE,
    SCOPE_IDENTITY_PROVE,
    SCOPE_IDENTITY_DELEGATE,
    SCOPE_PRESENCE_REPRESENT,
    SCOPE_TRANSACT_PURCHASE,
    SCOPE_TRANSACT_SELL,
    SCOPE_PAYMENTS_SEND,
    SCOPE_PAYMENTS_RECEIVE,
    SCOPE_PAYMENTS_AUTHORIZE,
    SCOPE_CONTRACT_READ,
    SCOPE_CONTRACT_SIGN,
    SCOPE_DATA_READ,
    SCOPE_DATA_WRITE,
    SCOPE_DATA_DELETE,
    SCOPE_DATA_EXPORT,
    SCOPE_DATA_SHARE,
    SCOPE_EXECUTE_TOOL,
    SCOPE_EXECUTE_CODE,
    SCOPE_GENERATE_CONTENT,
    SCOPE_GENERATE_DEEPFAKE,
    SCOPE_PHYSICAL_ENTER,
    SCOPE_PHYSICAL_EXIT,
    SCOPE_PHYSICAL_ACTUATE,
    SCOPE_PHYSICAL_MANIPULATE,
    SCOPE_ROBOT_OPERATE,
    SCOPE_ROBOT_MOVE,
    SCOPE_ROBOT_INTERACT,
    SCOPE_DRONE_FLY,
    SCOPE_DRONE_DELIVER,
    SCOPE_DRONE_CAPTURE,
    SCOPE_VEHICLE_OPERATE,
    SCOPE_VEHICLE_TRANSPORT,
    SCOPE_VEHICLE_CHARGE,
    SCOPE_INFRASTRUCTURE_MONITOR,
    SCOPE_INFRASTRUCTURE_CONTROL,
    SCOPE_INFRASTRUCTURE_ACCESS,
    SCOPE_ACTUATE_VALVE,
    SCOPE_ACTUATE_MOTOR,
    SCOPE_ACTUATE_SWITCH,
})

# Wildcards expand ONLY to non-sensitive scopes.
_SCOPE_WILDCARDS: dict[str, tuple[str, ...]] = {
    "meeting:*": (
        SCOPE_MEETING_ATTEND,
        SCOPE_MEETING_SPEAK,
        SCOPE_MEETING_VIDEO,
        SCOPE_MEETING_CHAT,
        SCOPE_MEETING_SHARE_SCREEN,
    ),
    "comms:message:*": (SCOPE_COMMS_MESSAGE_READ, SCOPE_COMMS_MESSAGE_SEND),
    "comms:email:*": (SCOPE_COMMS_EMAIL_READ, SCOPE_COMMS_EMAIL_SEND),
    "comms:*": (
        SCOPE_COMMS_MESSAGE_READ,
        SCOPE_COMMS_MESSAGE_SEND,
        SCOPE_COMMS_EMAIL_READ,
        SCOPE_COMMS_EMAIL_SEND,
        SCOPE_COMMS_CALENDAR_READ,
        SCOPE_COMMS_CALENDAR_WRITE,
    ),
    "transact:*": (SCOPE_TRANSACT_PURCHASE, SCOPE_TRANSACT_SELL),
    "payments:*": (SCOPE_PAYMENTS_SEND, SCOPE_PAYMENTS_RECEIVE),
    "data:*": (SCOPE_DATA_READ, SCOPE_DATA_SHARE),
    "execute:*": (SCOPE_EXECUTE_TOOL,),
    "generate:*": (SCOPE_GENERATE_CONTENT,),
    "physical:*": (SCOPE_PHYSICAL_ENTER, SCOPE_PHYSICAL_EXIT),
    "robot:*": (SCOPE_ROBOT_OPERATE, SCOPE_ROBOT_MOVE, SCOPE_ROBOT_INTERACT),
    "drone:*": (SCOPE_DRONE_DELIVER, SCOPE_DRONE_CAPTURE),
    "vehicle:*": (SCOPE_VEHICLE_TRANSPORT, SCOPE_VEHICLE_CHARGE),
    "infrastructure:*": (SCOPE_INFRASTRUCTURE_MONITOR,),
    # actuate:* — every member sensitive; NO wildcard expansion.
    # presence:* — presence:represent is sensitive, so NO wildcard expansion.
    # Representation must always be granted explicitly.
}


def _is_custom_scope(s: str) -> bool:
    return s.startswith(CUSTOM_SCOPE_PREFIX) and len(s) > len(CUSTOM_SCOPE_PREFIX)


def validate_scopes(scopes: list[str]) -> str | None:
    """Return an error message if any scope is invalid; None if all valid.

    Custom scopes (prefix "custom:") are accepted as valid extensions.
    """
    for s in scopes:
        if s in _VALID_SCOPES:
            continue
        if s in _SCOPE_WILDCARDS:
            continue
        if _is_custom_scope(s):
            continue
        return f'unknown scope "{s}": not in canonical vocabulary and not a custom: extension'
    return None


def is_sensitive(scope: str) -> bool:
    """True if the scope is flagged as sensitive. Custom scopes are
    non-sensitive by default; applications may enforce policy out-of-band."""
    return scope in _SENSITIVE_SCOPES


def expand_scopes(scopes: list[str]) -> list[str]:
    """Replace wildcards with their constituent non-sensitive scopes.

    Deduplicates and returns results lex-sorted for deterministic output.
    Custom scopes pass through unchanged.
    """
    seen: set[str] = set()
    for s in scopes:
        children = _SCOPE_WILDCARDS.get(s)
        if children is not None:
            seen.update(children)
        else:
            seen.add(s)
    return sorted(seen)


def has_scope(granted: list[str], required: str) -> bool:
    """True iff `required` is present after wildcard expansion of `granted`."""
    return required in expand_scopes(granted)


def intersect_scopes(*lists: list[str]) -> list[str]:
    """Set of scopes in every input list after wildcard expansion. Lex-sorted."""
    if not lists:
        return []
    effective = set(expand_scopes(lists[0]))
    for lst in lists[1:]:
        effective &= set(expand_scopes(lst))
    return sorted(effective)
